Add troubleshooting docs for atomic transactions (#1883)
vitessio/website: content/en/docs/21.0/user-guides/configuration-advanced/troubleshooting-atomic-transactions.md

---
title: Troubleshooting Distributed Atomic Transactions
weight: 65
---

> For users running sharded keyspaces in Vitess, we support atomic transactions across multiple shards using 2PC (Two-Phase Commit). You can find more details on configuring your cluster for distributed transactions in the [Distributed Transaction](../../../reference/features/distributed-transaction/) documentation.

Reviewer: ```suggestion
For users running sharded keyspaces in Vitess, we support atomic transactions across multiple shards using 2PC (Two-Phase Commit). You can find more details on configuring your cluster for distributed transactions in the [Distributed Transactions](../../../reference/features/distributed-transaction/) documentation.
```

Author: Fixed

Although distributed transactions are designed to run smoothly, issues may occasionally occur. This guide will help you troubleshoot and resolve these issues.

## Finding Issues

There are several methods to determine whether something has gone wrong in distributed transactions: 

1. **Monitor `vttablet` metrics** - The `vttablet` process exposes various metrics that can be used to monitor the health of the system. Here are the ones to look out for -
   1. `CommitPreparedFail` - This metric indicates the number of times a commit has failed after being prepared. It is categorized into two classes of errors: retryable and non-retryable. Non-retryable errors will require human intervention.
   2. `RedoPreparedFail` - Similar to `CommitPreparedFail`, this metric shows the number of times redoing a transaction has failed after preparation. Non-retryable errors here will also need human intervention.

> 2. **`GetUnresolvedTransactions` RPC** - This command lists unresolved transactions in the system. You can check the state of these transactions to determine if any are stuck in a failed state. More details on the command can be found [here](../../../reference/programs/vtctldclient/vtctldclient_distributedtransaction/vtctldclient_distributedtransaction_unresolved-list).

Reviewer: Instead of saying RPC, we should say CLI command.

Author: Fixed

> 2. **`GetUnresolvedTransactions` RPC** - This command lists unresolved transactions in the system. You can check the state of these transactions to determine if any are stuck in a failed state. More details on the command can be found [here](../../../reference/programs/vtctldclient/vtctldclient_distributedtransaction/vtctldclient_distributedtransaction_unresolved-list).

Reviewer: ```suggestion
2. **`GetUnresolvedTransactions`** - This CLI command lists unresolved transactions in the system. More details on the command can be found [here](../../../reference/programs/vtctldclient/vtctldclient_distributedtransaction/vtctldclient_distributedtransaction_unresolved-list).
```

Reviewer: I prefer to link from text instead of "more details....". You could use `[CLI command]` to link the reference page.

Author: Fixed

> 3. **Transactions page in VTAdmin** - The VTAdmin UI provides a page to view all unresolved transactions in the system. It uses the same RPC command as mentioned above.

Reviewer: ```suggestion
3. **Transactions page in VTAdmin** - The VTAdmin UI provides a page to view all unresolved transactions in the system. This view can be used to get the DTIDs for such transactions.
```

Author: Fixed

> ## Fixing The Issue

Reviewer: ```suggestion
## Fixing Issues
```

Author: Fixed

> Once a transaction has been identified as irreversibly failed, human intervention is required to correct the database state. Follow these steps:

Reviewer: ```suggestion
Once a transaction has been identified as irreversibly failed, human intervention is required to correct the database state. You will need the DTID(s) from the previous step. Follow these steps:
```

Author: Fixed

> 1. Query the `redo_state` and `redo_statement` tables on all participating shards using the dtid to get more information about the transaction.

Reviewer: ```suggestion
1. Query the `redo_state` and `redo_statement` tables on all participating shards using the DTID to get more information about the transaction.
```

Author: Fixed

> 1. Query the `redo_state` and `redo_statement` tables on all participating shards using the DTID to get more information about the transaction.

Reviewer 2: `redo_state` and `redo_statement` tables are not accessible directly to users, they have to connect directly to MySQL to retrieve this information. Probably mention this, and is created in _vt/sidecar db

Author: Created - https://github.com/vitessio/website/pull/1895

2. Each shard will display the status of the transaction on that shard. For shards where the transaction has failed, the information will include the writes that were part of the transaction.

> 3. Using this information, rerun the writes on all the failing shards.

Reviewer: ```suggestion
3. Using this information, rerun the relevant writes directly on each of the failing shards.
```

Author: Fixed

> 4. After making the necessary changes, conclude the transaction by running the `ConcludeTransaction` RPC command. More details about this command can be found [here](../../../reference/programs/vtctldclient/vtctldclient_distributedtransaction/vtctldclient_distributedtransaction_conclude).

Reviewer: ```suggestion
4. After making the necessary changes, conclude the transaction by running the `ConcludeTransaction` CLI command. More details about this command can be found [here](../../../reference/programs/vtctldclient/vtctldclient_distributedtransaction/vtctldclient_distributedtransaction_conclude).
```

Reviewer: Link from `[CLI command]` instead of saying "More details ...."

Author: Fixed

> Please open an issue on the Vitess GitHub repository to allow further investigation and resolution of the problem.

Reviewer: ```suggestion
If you do run into distributed transaction failures, please open a GitHub [issue] to allow further investigation and resolution of the problem.
```

Reviewer: Link `[issue]` to the `New Issue` page for vitessio/vitess.

Author: Fixed

---
title: Troubleshooting Distributed Atomic Transactions
weight: 65
---

For users running sharded keyspaces in Vitess, we support atomic transactions across multiple shards using 2PC (Two-Phase Commit). More details on how to configure your cluster to allow distributed transactions can be found in the [Distributed Transaction](../../../reference/features/distributed-transaction/) documentation.

> While we don't expect any problems, it is possible that you may encounter issues when running distributed transactions. This guide will help you troubleshoot and resolve these issues.

Reviewer 3: ```suggestion
Although distributed transactions are designed to run smoothly, issues may occasionally occur. This guide will help you troubleshoot and resolve these issues.
```

Author: Fixed

## Finding Issues

> There are a couple of ways to know whether something has gone wrong in distributed transactions - 

Reviewer 3: ```suggestion
There are several methods to determine whether something has gone wrong in distributed transactions:
```

Author: Fixed

> 1. **Metrics provided by `vttablet`s** - The `vttablet` process exposes a number of metrics that can be used to monitor the health of the system. Here are the ones to look out for -

Reviewer 3: ```suggestion
1. **Monitor `vttablet` metrics** - The `vttablet` process exposes various of metrics that can be used to monitor the health of the system. Here are the ones to look out for -
```

Author: Fixed

   1. `CommitPreparedFail` - This metric shows the number of times a commit has failed after it was already prepared. It is further sub-divided into 2 class of errors. One retryable, and the other not. Any non-retryable error will require human intervention.
   2. `RedoPreparedFail` - Similar to the previous metric, it shows the number of times a redoing a transaction has failed after it was already prepared. Any non-retryable error here too will require human intervention.

> 2. **`GetUnresolvedTransactions` RPC** - This command shows the list of unresolved transactions in the system. The state of these transactions can be checked to see if they are stuck in a failed state or not. More details on the command can be found [here](//TODO: add link).

Author: There are still TODO links in the doc

Author: Fixed

> 2. **`GetUnresolvedTransactions` RPC** - This command shows the list of unresolved transactions in the system. The state of these transactions can be checked to see if they are stuck in a failed state or not. More details on the command can be found [here](//TODO: add link).

Reviewer 3: ```suggestion
2. **`GetUnresolvedTransactions` RPC** - This command shows the list of unresolved transactions in the system. The state of these transactions can be checked to see if they are stuck in a failed state. More details on the command can be found [here](//TODO: add link).
```

Author: Fixed

3. **Transactions page in VTAdmin** - The VTAdmin UI provides a page to view all the unresolved transactions in the system. It works using the RPC same command as the previous one.

## Fixing The Issue

> Once we have found a transaction that is permanently failed, we need human intervention to fix the database state. Here are the steps to follow -

Reviewer 3: ```suggestion
Once we have found a transaction that is irreversibly failed, we need human intervention to fix the database state. Here are the steps to follow -
```

Author: Fixed

1. Get more information about the transaction by querying the `redo_state` and `redo_statement` tables in all the participating shards using the dtid.
2. For each shard, it will have the status of the transaction on that shard. For the shards where the transaction has failed, it will have the writes that were part of the transaction.
3. With this information, you can rerun the writes on all the failing shards.
4. Once all the changes are made, the transaction can be concluded by running the `ConcludeTransaction` RPC command. More information about the command can be found [here](//TODO: add link).

The same information can be gleaned from VTAdmin page by navigating to the transactions page, and clicking on the transaction in question.

> We would also request you to open an issue on Vitess, so that the underlying problem can be investigated and fixed.

Reviewer 3: ```suggestion
Please open an issue on the Vitess GitHub repository to allow further investigation and resolution of the problem.
```

Author: Fixed
